api/src/main/java/net/kyori/adventure/text/event/ClickCallback.java

```diff
@@ -0,0 +1,246 @@
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> filter(final @NotNull Predicate<T> filter) {
+    return this.filter(filter, null);
+  }
+
+  /**
+   * Filter audiences that receive this click callback.
+   *
+   * @param filter the filter to test audiences with
+   * @param failureMessage a message to send if the conditions are not met
+   * @return a filtered callback action
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> filter(final @NotNull Predicate<T> filter, final @Nullable Component failureMessage) {
+    return audience -> {
+      if (filter.test(audience)) {
+        this.accept(audience);
+      } else if (failureMessage != null) {
+        audience.sendMessage(failureMessage);
+      }
+    };
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always pass.</p>
+   *
+   * <p>Actions from audiences that do not match this predicate will be silently ignored.</p>
+   *
+   * @param permission the permission to check
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission) {
+    return this.requiringPermission(permission, null);
+  }
+
+  /**
+   * Require that audiences receiving this callback have a certain permission.
+   *
+   * <p>For audiences without permissions information, this test will always pass.</p>
+   *
+   * @param permission the permission to check
+   * @param failureMessage a message to send if the conditions are not met
+   * @return a modified callback
+   * @since 4.13.0
+   */
+  default @NotNull ClickCallback<T> requiringPermission(final @NotNull String permission, final @Nullable Component failureMessage) {
+    return this.filter(audience -> audience.getOrDefault(PermissionChecker.POINTER, PermissionChecker.always(TriState.TRUE)).test(permission), failureMessage);
+  }
+
+  /**
+   * Options to configure how a callback can be executed.
+   *
+   * @since 4.13.0
+   */
+  @ApiStatus.NonExtendable
+  interface Options extends Examinable {
+    /**
+     * Create a new builder.
+     *
+     * @return the new builder
+     * @since 4.13.0
+     */
+    static @NotNull Builder builder() {
+      return new ClickCallbackOptionsImpl.BuilderImpl();
+    }
+
+    /**
+     * Create a new builder populating from existing options.
+     *
+     * @param existing the existing options to populate this builder with
+     * @return the new builder
+     * @since 4.13.0
+     */
+    static @NotNull Builder builder(final @NotNull Options existing) {
+      return new ClickCallbackOptionsImpl.BuilderImpl(existing);
+    }
+
+    /**
+     * Whether the callback can be executed multiple times.
+     *
+     * <p>By default callbacks are single-use.</p>
+     *
+     * @return whether this callback is multi-use
+     * @since 4.13.0
+     */
+    boolean multiUse();
```

Comment:
```suggestion
    int uses();
```

Reply:
Maybe? Are there many use cases for use counts between 1 and "infinite"?

Reply:
could be? just seems a bit friendlier to allow the user of the library to determine how many uses a callback can have?

Reply:
thinking of stuff like "first five people to click this message gets some diamonds"